File: rock/node-cli/commands/model_service.js
const fs = require('fs-extra');
const path = require('path');

const MODEL_SERVICE_DIR = 'data/cli/model';
const PID_FILE = path.join(MODEL_SERVICE_DIR, 'pid.txt');

const modelServiceCommand = {
  command: 'model-service command',
  describe: 'Model service operations',
  builder: (yargs) => {
    return yargs
      .command(
        'start',
        'Start model service',
        (yargs) => {
          return yargs.option('type', {
            describe: 'Type of model service',
            type: 'string',
            choices: ['local', 'proxy'],
            default: 'local'
          });
        },
        async (argv) => {
          await startModelService(argv);
        }
      )
      .command(
        'watch-agent',
        'Watch agent status, if stopped, send SESSION_END',
        (yargs) => {
          return yargs.option('pid', {
            describe: 'PID of agent process to watch',
            type: 'number',
            demandOption: true
          });
        },
        async (argv) => {
          await watchAgent(argv);
        }
      )
      .command(
        'stop',
        'Stop model service',
        () => { },
        async () => {
          await stopModelService();
        }
      )
      .command(
        'anti-call-llm',
        'Anti call llm, input is response of llm, output is the next request to llm',
        (yargs) => {
          return yargs
            .option('index', {
              describe: 'Index of last llm call, start from 0',
              type: 'number',
              demandOption: true
            })
            .option('response', {
              describe: 'Response of last llm call',
              type: 'string'
            });
        },
        async (argv) => {
          await antiCallLlm(argv);
        }
      );
  },
  handler: async () => {
    // This will be handled by subcommands
  }
};

const startModelService = async (argv) => {
  // Create model service directory if it doesn't exist
  await fs.ensureDir(MODEL_SERVICE_DIR);

  // Check if model service already exists
  if (await modelServiceExists()) {
    console.error('Model service already exists, please run \'rock model-service stop\' first');
    process.exit(1);
    return;
  }

  console.log('Starting model service...');
  // Implementation would go here
  // This would call the actual model service start functionality
  const serviceType = argv.type;
  console.log(`Starting model service with type: ${serviceType}`);

  // In a real implementation, this would start the actual service and return a PID
  const mockPid = Math.floor(Math.random() * 10000) + 1000; // Mock PID for demonstration
  console.log(`Model service started, PID: ${mockPid}`);

  // Write PID to file
  await fs.writeFile(PID_FILE, mockPid.toString());
};

const watchAgent = async (argv) => {
  const agentPid = argv.pid;
  console.log(`Starting to watch agent process, PID: ${agentPid}`);
  // Implementation would go here
  // This would call the actual watch agent functionality
};

const stopModelService = async () => {
  if (!(await modelServiceExists())) {
    console.log('Model service does not exist, skipping');
    return;
  }

  console.log('Stopping model service...');
  // Read PID from file
  const pid = await fs.readFile(PID_FILE, 'utf8');
  console.log(`Model service exists, PID: ${pid}`);

  // Implementation would go here
  // This would call the actual model service stop functionality
  try {
    process.kill(parseInt(pid), 'SIGTERM');
    console.log(`Model service with PID ${pid} terminated`);
  } catch (e) {
    console.error(`Could not stop process with PID ${pid}: ${e.message}`);
  }

  // Remove PID file
  await fs.remove(PID_FILE);
  console.log('Model service stopped');
};

const antiCallLlm = async (argv) => {
  console.log('Starting anti call llm...');
  // Implementation would go here
  // This would call the actual anti call llm functionality
  const index = argv.index;
  const response = argv.response;

  console.log(`Index: ${index}`);
  if (response) {
    console.log(`Response: ${response}`);
  }

  // In a real implementation, this would return the next request
  const nextRequest = `Next request for call #${index}`;
  console.log(nextRequest); // This is what gets printed to stdout
};

const modelServiceExists = async () => {
  const exists = await fs.pathExists(PID_FILE);
  if (exists) {
    const pid = await fs.readFile(PID_FILE, 'utf8');
    console.log(`Model service exists, PID: ${pid}`);
  }
  return exists;
};

module.exports = modelServiceCommand;

File: rock/cli/command/model_service.py
import argparse
import logging
import sys
from pathlib import Path

from rock.cli.command.command import Command
from rock.sdk.model.client import ModelClient
from rock.sdk.model.service import ModelService

logger = logging.getLogger(__name__)


class ModelServiceCommand(Command):
    """
    Command for managing the model service.

    Provides subcommands to start, stop, and monitor the model service,
    which can run in 'local' mode (using a sandboxed local LLM) or 'proxy'
    mode (forwarding requests to an external endpoint).

    Examples:
        # Start local model service on default port 8080
        rock model-service start --type local

        # Start with custom port
        rock model-service start --type local --port 9000

        # Start with a configuration file
        rock model-service start --type local --config-file config.yaml

        # Start proxy mode with external endpoint
        rock model-service start --type proxy --proxy-url https://api.openai.com/v1

        # Start with custom retry behavior
        rock model-service start --type proxy --retryable-status-codes 429,500,502

        # Stop the running service
        rock model-service stop
    """

    name = "model-service"

    DEFAULT_MODEL_SERVICE_DIR = "data/cli/model"
    DEFAULT_MODEL_SERVICE_PID_FILE = DEFAULT_MODEL_SERVICE_DIR + "/pid.txt"

    async def arun(self, args: argparse.Namespace):
        if not Path(self.DEFAULT_MODEL_SERVICE_DIR).exists():
            Path(self.DEFAULT_MODEL_SERVICE_DIR).mkdir(parents=True, exist_ok=True)

        sub_command = args.model_service_command
        if "start" == sub_command:
            if await self._model_service_exist():
                logger.error("model service already exist, please run 'rock model-service stop' first")
                sys.exit(1)
                return
            logger.info("start model service")
            model_service = ModelService()
            pid = await model_service.start(
                model_service_type=args.type,
                config_file=args.config_file,
                host=args.host,
                port=args.port,
                proxy_url=args.proxy_url,
                retryable_status_codes=args.retryable_status_codes,
                request_timeout=args.request_timeout,
            )
            logger.info(f"model service started, pid: {pid}")
            with open(self.DEFAULT_MODEL_SERVICE_PID_FILE, "w") as f:
                f.write(pid)
            return
        if "watch-agent" == sub_command:
            agent_pid = args.pid
            logger.info(f"start to watch agent process, pid: {agent_pid}")
            model_service = ModelService()
            await model_service.start_watch_agent(agent_pid)
            return
        if "stop" == sub_command:
            if not await self._model_service_exist():
                logger.info("model service not exist, skip")
                return
            logger.info("start to stop model service")
            with open(self.DEFAULT_MODEL_SERVICE_PID_FILE) as f:
                pid = f.read()
                model_service = ModelService()
                await model_service.stop(pid)
            Path(self.DEFAULT_MODEL_SERVICE_PID_FILE).unlink()
            logger.info("model service stopped")
            return
        if "anti-call-llm" == sub_command:
            logger.debug("start to anti call llm")
            model_client = ModelClient()
            next_request = await model_client.anti_call_llm(index=args.index, last_response=args.response)
            # necessary: print next_request to stdout, and do NOT print anything else
            print(next_request)

    async def _model_service_exist(self) -> bool:
        exist = Path(self.DEFAULT_MODEL_SERVICE_PID_FILE).exists()
        if exist:
            with open(self.DEFAULT_MODEL_SERVICE_PID_FILE) as f:
                pid = f.read()
                logger.info(f"model service exist, pid: {pid}.")
        return exist

    @staticmethod
    async def add_parser_to(subparsers: argparse._SubParsersAction):
        model_service_parser = subparsers.add_parser(
            "model-service",
            description="model-service command",
        )
        model_service_subparsers = model_service_parser.add_subparsers(
            dest="model_service_command",
        )

        # rock model-service start
        start_parser = model_service_subparsers.add_parser(
            "start",
            help="start model service",
        )
        start_parser.add_argument(
            "--type",
            type=str,
            choices=["local", "proxy"],
            default="local",
            help="Type of model service (local/proxy)",
        )
        start_parser.add_argument(
            "--config-file",
            type=str,
            default=None,
            help="Path to the configuration YAML file",
        )
        start_parser.add_argument(
            "--host",
            type=str,
            default=None,
            help="Server host address. Overrides config file.",
        )
        start_parser.add_argument(
            "--port",
            type=int,
            default=None,
            help="Server port. Overrides config file.",
        )
        start_parser.add_argument(
            "--proxy-url",
            type=str,
            default=None,
            help="Direct proxy URL (e.g., https://your-endpoint.com/v1). Takes precedence over config file.",
        )
        start_parser.add_argument(
            "--retryable-status-codes",
            type=str,
            default=None,
            help="Retryable status codes, comma-separated (e.g., '429,500,502'). Overrides config file.",
        )
        start_parser.add_argument(
            "--request-timeout",
            type=int,
            default=None,
            help="Request timeout in seconds. Overrides config file.",
        )

        watch_agent_parser = model_service_subparsers.add_parser(
            "watch-agent",
            help="watch agent status, if stopped, send SESSION_END",
        )
        watch_agent_parser.add_argument(
            "--pid",
            required=True,
            type=int,
            help="pid of agent process to watch",
        )

        # rock model-service stop
        model_service_subparsers.add_parser(
            "stop",
            help="stop model service",
        )

        # rock model-service anti-call-llm --index N [--response RESPONSE]
        anti_call_llm_parser = model_service_subparsers.add_parser(
            "anti-call-llm",
            help="anti call llm, input is response of llm, output is the next request to llm",
        )
        anti_call_llm_parser.add_argument(
            "--index", required=True, type=int, help="index of last llm call, start from 0"
        )
        anti_call_llm_parser.add_argument("--response", required=False, help="response of last llm call")
